### ISSUE TITLE
Safe Rust API Conventions & Implementation Strategy

### ISSUE THREAD

**Issue by pragmatrix** [rust-skia/rust-skia#24]
I finally got around to port the canvas example to a _safe API_ according to my limited knowledge of Rust. 

What I learned so far is, that the bindgen bindings are not always usable, specifically when C++ classes, like smartpointers `sk_sp` are returned (or passed?) by value, the calling conventions break apart, so all C++ methods that utilize that, need to be wrapped before the can be used in Rust.

This ticket should clarify the implementation strategy for the safe Rust API and eventually form a wiki page.

## C Naming conventions

- The `extern "C"` functions that we need to add to the bindings should begin with `extern "C" C_`, this way we can whitelist them all with one match pattern `C_.*` in bindgen. They include the class / struct name they are interfacing with and then the method's name. Name parts are separated by `_`.

## Rust naming conventions

I would like to avoid any abbrevations that are appearing in struct names, like `Sk` or something like that. For the skia binding Surface, for example, I would prefer to use the name `skia::Surface`. A `GrContext` should be named `GraphicsContext` or `graphics::Context`. 

Static "constructor" functions that begin in Skia with `Make` should begin with `new_` in Rust, abbrevations that indicate function parameter variants may be used (`Rect::new_iwh()`, (i)int (w)idth (h)eight for example) . In general, the naming should be _very similar_ to the original API but when possible be Rust idiomatic, so that a Rust user can use the original Skia documentation without much mental translation.

## Smartpointers / Ref counting base classes

- Classes with reference counters that need to be used in Rust are transported as native pointers, the function `.release()` can be used for that.
- If a pointer derives from a smartpointer base class, like `SkRefCntBase` or `SkNVRefCnt`, the `unref()` function has to be called in the related `Drop` trait. If we need to use smart pointers in Rust, we can interface with `ref()` and `unref()` later on, although I would like to avoid that at first and see how far we can go with Rust's ownership system.

For example:

```c++
extern "C" SkSurface* C_SkSurface_MakeRasterN32Premul(int width, int height, const SkSurfaceProps* surfaceProps) {
    return SkSurface::MakeRasterN32Premul(width, height, surfaceProps).release();
}
```

In Rust:

```rust
pub struct Surface {
    native: *mut SkSurface
}

impl Drop for Surface {
    fn drop(&mut self) {
        unsafe { (*self.native)._base._base.unref() }
    }
}

impl Surface {
    pub fn new_raster_n32_premul(width: i32, height: i32) -> Surface {
        Surface {
            native: unsafe { C_SkSurface_MakeRasterN32Premul(width, height, ptr::null()) }
        }
    }
}

```
## C++ Structures (like SkPaint and SkPath) for which the constructor / destructor _must_ be called.

Here bindgen generates `new()` and `destruct()` functions. If possible, I would prefer not to use pointers for these in Rust. From what I can see now, the `new()` function can be called from Rust, but I had no luck linking the `destruct()` functions. May be this is an error in bindgen or some tree shaking process removes them. So to fix that we can interface a destructor like this:

```c++
extern "C" void C_SkPaint_destruct(const SkPaint* self) {
    self->~SkPaint();
}
```

## Overloaded Functions

A lot of Skia functions offer overloads with different abstraction levels, so for example, a `draw_rect` function may exist with an overloaded function that takes

- 4 scalar values.
- 2 points.
- a `Rect` struct.

Since Rust does not support overloading functions, I see the following options:

- We support each of them, but name the function differently.

- We support a rich set of `.into` traits that convert more granular types into more abstract types. For example a `Rect` supports the following `From` traits (implicitly providing the `.into()` function):

  - `From<(f32, f32, f32, f32)>` 
  - `From<(Point, Point)>`

  So a `draw_rect` can be called with `draw_rect((x1, y1, x2, y2).into())`, or `draw_rect((p1, p2).into())` or just `draw_rect(rect)`.

  ... I am not sure if this even works, or if that is even a Rust idiomatic way to support something like overloaded functions.

- Use enums for the variants?

## Enumeration constants.

TBD

## Default parameters.

TBD: Option, Builder pattern?

**Comment by pragmatrix** [comment]
Following up on this, @M-Adoo What do you think now? I've added some examples and there were no stability issues so far. Is this library usable for you, and if not, what can we improve?

**Comment by pragmatrix** [comment]
I've added some of the "API Conventions & Interfacing Patterns" to the wiki. If something does not fit, please open a new issue. 

In case of typos, fixing the grammar, or other small changes, I think it's best to update the wiki content directly.

https://github.com/rust-skia/rust-skia/wiki

**Comment by M-Adoo** [comment]
Yes, I'll use it. You do a really good job. After support wasm, I'll use it  in my lib. Is it wasm support also in your plan?  If not, that's no problem. I'll try to add wasm support in next week.  I'm busy designing an another project recently, after that I'll do some contributes  to this repo.

**Comment by pragmatrix** [comment]
Thank you for the feedback!

I haven't planned to support WebAssembly, but there was originally a WebAssembly target configured in the CI script, so may be @Brooooooklyn can say more about that.

**Comment by Brooooooklyn** [comment]
Support wasm and nodejs native binding is my original purpose to start this project.
So yes, we will always consider about wasm support.
And wasm support has no problem for now, since we have checked it in pipelines.

**Comment by pragmatrix** [comment]
@Brooooooklyn I have removed that because it was only building the bindings and not skia AFAIK.

**Comment by pragmatrix** [comment]
Closed in favor of #39.
